File: programmers/algorithm_data_structure/heap/디스크 콘트롤러.py
import heapq as hq
from collections import deque


def my_solution(jobs):
    '''
    매번 대기중인 작업 중에서 실행 시간이 가장 작은 작업을
    heap을 활용해 고르는 게 중요

    [설명]
    A -> C -> B
    A는 0초 기다림
    C는 (A초 실행시간 - 요청 시작 지점) 만큼 기다림
    B는 (A초 실행시간 + C초 실행시간 - 요청 시작 지점) 만큼 기다림

    (요청 시작 지점은 상수이므로) 각 작업 실행 시점까지 총 작업 시간을
    최소화하며 가야 최소 평균 시간을 보장한다. (dynamic programming)

    [주의점]
    * 요청 시간이 각 시점에서 이미 지나야 선택 항목으로 쓸 수 있음
    * jobs는 정렬되어 있지 않아 시간 + 처리시간 순으로 정렬 필요
    * 작업을 끝냈을 때 다음 작업까지 실행 없이 대기하는 경우 고려
    '''
    answer = 0
    curr_time = 0
    sorted_jobs = deque(sorted(jobs))

    delayed_jobs = []

    # 모든 작업이 완료될 때까지 반복
    while sorted_jobs or delayed_jobs:
        # 우선순위에 따라 수행할 작업 선택
        job = None
        request_time = None
        processing_time = None

        if delayed_jobs:
            job = hq.heappop(delayed_jobs)
            processing_time, request_time = job
        else:
            job = sorted_jobs.popleft()
            request_time, processing_time = job

        curr_time += max(request_time - curr_time, 0) + processing_time
        answer += (curr_time - request_time)

        # 대기 작업 목록 업데이트
        while sorted_jobs and sorted_jobs[0][0] < curr_time:
            job = sorted_jobs.popleft()
            hq.heappush(delayed_jobs, (job[1], job[0]))

    return answer // len(jobs)

File: programmers/algorithm_data_structure/heap/디스크_콘트롤러.java
package algorithm_data_structure.heap;

import java.util.*;

class 디스크_콘트롤러 {
    public int solution(int[][] jobs) {
        /*
         * 0. 가장 먼저 들어오는 작업(시각이 같으면 처리 시간이 짧은 작업) 순으로 jobs를 정렬
         *    (문제에서 jobs 데이터가 먼저 들어오는 순으로 정렬되었다는 보장이 없음)
         * 1. 밀린 작업이 없다면 가장 먼저 들어오는 작업(시각이 같으면 처리 시간이 작은 것)을 처리
         * 2. 밀린 작업이 있다면 그중 작업 시간이 가장 짧은 것 처리
         * 
         * 작업 시간이 가장 짧아야 남은 작업들의 대기 시간이 최소화된다.
         */
        int timeSum = 0;
        int currTime = 0;
        PriorityQueue<int[]> sortedJobs = new PriorityQueue<>(
            (a1, a2) -> (a1[0] != a2[0]) ? (a1[0] - a2[0]) : (a1[1] - a2[1])
        );
        PriorityQueue<int[]> delayedJobs = new PriorityQueue<>((a1, a2) -> a1[1] - a2[1]);

        for (int[] job : jobs){
            sortedJobs.add(job);
        }

        while (!delayedJobs.isEmpty() || !sortedJobs.isEmpty()){
            int[] job = delayedJobs.isEmpty() ? sortedJobs.poll() : delayedJobs.poll();
            int requestTime = job[0];
            int processingTime = job[1];

            currTime += (Math.max(requestTime - currTime, 0) + processingTime);
            timeSum += (currTime - requestTime);

            while (!sortedJobs.isEmpty() && sortedJobs.peek()[0] < currTime) {
                delayedJobs.add(sortedJobs.poll());
            }
        }

        return (int) (1.0 * timeSum / jobs.length);
    }
}